File: go-version/internal/handlers/process.go
package handlers

import (
	"fmt"
	"io/fs"
	"net/http"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"file-processor/internal/logger"
	"file-processor/internal/models"
	"file-processor/internal/services"

	"github.com/gin-gonic/gin"
	"github.com/sirupsen/logrus"
)

type ProcessHandler struct {
	fileProcessor *services.FileProcessor
	textSanitizer *services.TextSanitizer
}

func NewProcessHandler(fileProcessor *services.FileProcessor, textSanitizer *services.TextSanitizer) *ProcessHandler {
	return &ProcessHandler{
		fileProcessor: fileProcessor,
		textSanitizer: textSanitizer,
	}
}

func (h *ProcessHandler) ProcessMessage(c *gin.Context) {
	var request models.ProcessMessageRequest
	if err := c.ShouldBindJSON(&request); err != nil {
		logger.WithFields(logrus.Fields{
			"error": err.Error(),
		}).Error("Validation error for /process-message")
		c.JSON(http.StatusBadRequest, gin.H{
			"error":   "Invalid request body",
			"details": err.Error(),
		})
		return
	}

	logger.WithFields(logrus.Fields{
		"messageCount": len(request),
	}).Info("Received /process-message request")

	if len(request) == 0 {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Request body must contain at least one message.",
		})
		return
	}

	// Processar arquivos de forma concorrente
	var allExtractedText strings.Builder
	var processedFiles []string
	var failedFiles []models.FailedFile
	var mutex sync.Mutex
	var wg sync.WaitGroup

	for _, message := range request {
		if message.Body.Files == nil || len(message.Body.Files) == 0 {
			continue
		}

		for _, file := range message.Body.Files {
			wg.Add(1)
			go func(file models.FileInfo) {
				defer wg.Done()

				textContent, err := h.fileProcessor.ProcessFile(file)
				if err != nil {
					logger.WithFields(logrus.Fields{
						"fileId": file.FileID,
						"error":  err.Error(),
					}).Error("Failed to process file")

					mutex.Lock()
					failedFiles = append(failedFiles, models.FailedFile{
						FileID: file.FileID,
						Error:  err.Error(),
					})
					mutex.Unlock()
					return
				}

				// Extrair nome do arquivo da URL
				parsedURL, err := url.Parse(file.URL)
				if err != nil {
					parsedURL = &url.URL{Path: file.FileID}
				}
				fileName := path.Base(parsedURL.Path)
				header := fmt.Sprintf("## Transcricao do arquivo: %s:\n\n", fileName)
				result := header + textContent

				mutex.Lock()
				processedFiles = append(processedFiles, file.FileID)
				allExtractedText.WriteString(result)
				allExtractedText.WriteString("\n\n---\n\n")
				mutex.Unlock()
			}(file)
		}
	}

	// Aguardar todos os processamentos
	wg.Wait()

	// Criar diretório de textos se não existir
	textsDir := "./public/texts"
	if err := os.MkdirAll(textsDir, 0755); err != nil {
		logger.WithFields(logrus.Fields{
			"error": err.Error(),
		}).Error("Failed to create texts directory")
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Internal server error",
			"message": "Failed to create output directory",
		})
		return
	}

	// Gerar nome do arquivo
	conversationID := request[0].ConversationID
	filename := fmt.Sprintf("%s-%d.txt", conversationID, time.Now().UnixMilli())
	filePath := filepath.Join(textsDir, filename)

	// Sanitizar texto final antes de salvar
	finalText := strings.TrimSpace(allExtractedText.String())
	sanitizedText := h.textSanitizer.SanitizeText(finalText)

	// Salvar arquivo
	if err := os.WriteFile(filePath, []byte(sanitizedText), 0644); err != nil {
		logger.WithFields(logrus.Fields{
			"error":    err.Error(),
			"filePath": filePath,
		}).Error("Failed to write text file")
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Internal server error",
			"message": "Failed to save processed text",
		})
		return
	}

	// Construir URL de download
	scheme := "http"
	if c.Request.TLS != nil {
		scheme = "https"
	}
	downloadURL := fmt.Sprintf("%s://%s/texts/%s", scheme, c.Request.Host, filename)

	logger.WithFields(logrus.Fields{
		"conversationId": conversationID,
		"downloadUrl":    downloadURL,
	}).Info("Successfully processed messages and created text file")

	c.JSON(http.StatusOK, models.ProcessMessageResponse{
		ConversationID: conversationID,
		DownloadURL:    downloadURL,
		ProcessedFiles: processedFiles,
		FailedFiles:    failedFiles,
	})
}

func (h *ProcessHandler) DeleteTexts(c *gin.Context) {
	var request models.DeleteTextsRequest
	if err := c.ShouldBindJSON(&request); err != nil {
		// Se não há body, aceita parâmetros vazios
		request = models.DeleteTextsRequest{}
	}

	logger.WithFields(logrus.Fields{
		"filename":       request.Filename,
		"conversationId": request.ConversationID,
	}).Info("Received /delete-texts request")

	textsDir := "./public/texts"

	// Verificar se o diretório existe
	if _, err := os.Stat(textsDir); os.IsNotExist(err) {
		logger.WithFields(logrus.Fields{
			"textsDir": textsDir,
		}).Warn("Texts directory does not exist")
		c.JSON(http.StatusNotFound, models.DeleteTextsResponse{
			Message:      "Texts directory not found",
			DeletedFiles: []string{},
			DeletedCount: 0,
		})
		return
	}

	// Ler todos os arquivos do diretório
	files, err := os.ReadDir(textsDir)
	if err != nil {
		logger.WithFields(logrus.Fields{
			"error": err.Error(),
		}).Error("Failed to read texts directory")
		c.JSON(http.StatusInternalServerError, gin.H{
			"error":   "Internal server error",
			"message": "Failed to read texts directory",
		})
		return
	}

	// Filtrar apenas arquivos .txt
	var txtFiles []fs.DirEntry
	for _, file := range files {
		if !file.IsDir() && strings.HasSuffix(file.Name(), ".txt") {
			txtFiles = append(txtFiles, file)
		}
	}

	if len(txtFiles) == 0 {
		logger.Info("No txt files found in texts directory")
		c.JSON(http.StatusNotFound, models.DeleteTextsResponse{
			Message:      "No txt files found to delete",
			DeletedFiles: []string{},
			DeletedCount: 0,
		})
		return
	}

	// Determinar quais arquivos excluir
	var filesToDelete []string

	if request.Filename != "" {
		// Excluir arquivo específico
		found := false
		for _, file := range txtFiles {
			if file.Name() == request.Filename {
				filesToDelete = append(filesToDelete, file.Name())
				found = true
				break
			}
		}
		if !found {
			logger.WithFields(logrus.Fields{
				"filename": request.Filename,
			}).Warn("Specific file not found")
			c.JSON(http.StatusNotFound, models.DeleteTextsResponse{
				Message:      fmt.Sprintf("File %s not found", request.Filename),
				DeletedFiles: []string{},
				DeletedCount: 0,
			})
			return
		}
	} else if request.ConversationID != "" {
		// Excluir arquivos de uma conversa específica
		for _, file := range txtFiles {
			if strings.HasPrefix(file.Name(), request.ConversationID) {
				filesToDelete = append(filesToDelete, file.Name())
			}
		}
		if len(filesToDelete) == 0 {
			logger.WithFields(logrus.Fields{
				"conversationId": request.ConversationID,
			}).Warn("No files found for conversation")
			c.JSON(http.StatusNotFound, models.DeleteTextsResponse{
				Message:      fmt.Sprintf("No files found for conversation %s", request.ConversationID),
				DeletedFiles: []string{},
				DeletedCount: 0,
			})
			return
		}
	} else {
		// Excluir todos os arquivos txt
		for _, file := range txtFiles {
			filesToDelete = append(filesToDelete, file.Name())
		}
	}

	// Excluir os arquivos
	var deletedFiles []string
	var failedFiles []models.FailedFile

	for _, filename := range filesToDelete {
		filePath := filepath.Join(textsDir, filename)
		if err := os.Remove(filePath); err != nil {
			logger.WithFields(logrus.Fields{
				"file":  filename,
				"error": err.Error(),
			}).Error("Failed to delete file")
			failedFiles = append(failedFiles, models.FailedFile{
				FileID: filename,
				Error:  err.Error(),
			})
		} else {
			deletedFiles = append(deletedFiles, filename)
			logger.WithFields(logrus.Fields{
				"file": filename,
			}).Info("File deleted successfully")
		}
	}

	response := models.DeleteTextsResponse{
		Message:      fmt.Sprintf("Successfully deleted %d file(s)", len(deletedFiles)),
		DeletedFiles: deletedFiles,
		DeletedCount: len(deletedFiles),
	}

	if len(failedFiles) > 0 {
		response.FailedFiles = failedFiles
		response.Message += fmt.Sprintf(". Failed to delete %d file(s)", len(failedFiles))
	}

	logger.WithFields(logrus.Fields{
		"deletedCount": len(deletedFiles),
		"failedCount":  len(failedFiles),
	}).Info("Delete operation completed")

	c.JSON(http.StatusOK, response)
}

File: src/routes/process.ts
import express, { type Request, type Response } from 'express';
import fs from 'node:fs';
import path from 'node:path';
import { z, ZodError } from 'zod';
import logger from '../lib/logger';
import { processDocx, processImage, processPdf, processTxt } from '../services/fileProcessor';
import { sanitizeText } from '../utils/textSanitizer';
const router = express.Router();


const FileInfoSchema = z.object({
  fileId: z.string(),
  url: z.string().url(),
  mimeType: z.string(),
  fileType: z.enum(['txt', 'pdf', 'jpeg', 'png', 'jpg', 'docx', 'image']),
}).loose();

const BodySchema = z.object({
  content: z.string().optional(),
  files: z.array(FileInfoSchema).optional(),
});

const MessageSchema = z.object({
  conversationId: z.string(),
  body: BodySchema
}).passthrough();

const ProcessMessageSchema = z.array(MessageSchema);

/**
 * @openapi
 * /process-message:
 *   post:
 *     summary: Processa uma ou mais mensagens com texto e arquivos, enviando o conteúdo para o Redis
 *     tags:
 *       - Processamento
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: array
 *             items:
 *               type: object
 *               properties:
 *                 conversationId:
 *                   type: string
 *                   description: ID único para a conversa/sessão.
 *                 body:
 *                   type: object
 *                   properties:
 *                     content:
 *                       type: string
 *                       description: Conteúdo textual opcional a ser processado.
 *                     files:
 *                       type: array
 *                       items:
 *                         type: object
 *                         properties:
 *                           fileId:
 *                             type: string
 *                           url:
 *                             type: string
 *                             format: uri
 *                           mimeType:
 *                             type: string
 *                           fileType:
 *                             type: string
 *                             enum: ['txt', 'pdf', 'jpeg', 'png', 'jpg', 'docx', 'image']
 *                         required:
 *                           - fileId
 *                           - url
 *                           - mimeType
 *                           - fileType
 *     responses:
 *       '200':
 *         description: Processamento concluído e URL para download do texto concatenado.
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 conversationId:
 *                   type: string
 *                   description: O ID da conversa da primeira mensagem processada.
 *                 downloadUrl:
 *                   type: string
 *                   format: uri
 *                   description: A URL pública para baixar o arquivo .txt com todo o conteúdo de texto extraído.
 *                 processedFiles:
 *                   type: array
 *                   items:
 *                     type: string
 *                 failedFiles:
 *                   type: array
 *                   items:
 *                     type: object
 *       '400':
 *         description: Erro de validação nos dados de entrada.
 *       '500':
 *         description: Erro interno do servidor.
 */
router.post('/process-message', async (req: Request, res: Response) => {
  try {
    const rawMessages = Array.isArray(req.body) ? req.body : [req.body];
    const messages = ProcessMessageSchema.parse(rawMessages);

    logger.info('Received /process-message request', { messageCount: messages.length });

    if (messages.length === 0) {
      return res.status(400).json({ error: 'Request body must contain at least one message.' });
    }

    let allExtractedText = '';
    const processedFiles: string[] = [];
    const failedFiles: { fileId: string; error: string }[] = [];

    for (const message of messages) {
      const { body } = message;
      const { files } = body;

      if (files && files.length > 0) {
        const processingPromises = files.map(async (file) => {
          try {
            let textContent = '';
            switch (file.fileType) {
              case 'txt':
                textContent = await processTxt(file);
                break;
              case 'pdf':
                textContent = await processPdf(file);
                break;
              case 'jpeg':
              case 'jpg':
              case 'png':
              case 'image':
                textContent = await processImage(file);
                break;
              case 'docx':
                textContent = await processDocx(file);
                break;
              default:
                logger.warn('Unsupported file type, skipping', { fileType: file.fileType, fileId: file.fileId });
                throw new Error('Unsupported file type');
            }
            processedFiles.push(file.fileId);
            const fileName = path.basename(new URL(file.url).pathname);
            const header = `## Transcricao do arquivo: ${fileName}:\n\n`;
            return header + textContent;
          } catch (error: any) {
            logger.error('Failed to process file', { fileId: file.fileId, error: error.message });
            failedFiles.push({ fileId: file.fileId, error: error.message });
            return null;
          }
        });

        const results = await Promise.all(processingPromises);
        results.forEach(text => {
          if (text) {
            allExtractedText += text + '\n\n---\n\n';
          }
        });
      }
    }

    const conversationId = messages[0].conversationId;
    const filename = `${conversationId}-${Date.now()}.txt`;
    const textsDir = path.join(__dirname, '..', '..', 'public', 'texts');

    fs.mkdirSync(textsDir, { recursive: true });

    const filePath = path.join(textsDir, filename);

    const sanitizedText = sanitizeText(allExtractedText.trim());
    fs.writeFileSync(filePath, sanitizedText);

    const downloadUrl = `${req.protocol}://${req.get('host')}/texts/${filename}`;

    logger.info('Successfully processed messages and created text file', { conversationId, downloadUrl });

    return res.status(200).json({
      conversationId,
      downloadUrl,
      processedFiles,
      failedFiles,
    });

  } catch (error: any) {
    if (error instanceof ZodError) {
      logger.error('Validation error for /process-message', { errors: error.issues });
      return res.status(400).json({ error: 'Invalid request body', details: error.issues });
    }
    logger.error('Error in /process-message handler', { error: error.message });
    return res.status(500).json({ error: 'Internal server error', message: error.message });
  }
});

interface DeleteTextsBody {
  filename?: string;
  conversationId?: string;
}

interface FailedFile {
  file: string;
  error: string;
}

router.delete('/delete-texts', async (req: Request, res: Response) => {
  try {
    const { filename, conversationId }: DeleteTextsBody = req.body || {};

    logger.info('Received /delete-texts request', { filename, conversationId });

    const textsDir = path.join(__dirname, '..', '..', 'public', 'texts');

    if (!fs.existsSync(textsDir)) {
      logger.warn('Texts directory does not exist', { textsDir });
      return res.status(404).json({
        error: 'Texts directory not found',
        deletedFiles: [],
        deletedCount: 0
      });
    }

    const allFiles = fs.readdirSync(textsDir);
    const txtFiles = allFiles.filter(file => file.endsWith('.txt'));

    if (txtFiles.length === 0) {
      logger.info('No txt files found in texts directory');
      return res.status(404).json({
        message: 'No txt files found to delete',
        deletedFiles: [],
        deletedCount: 0
      });
    }

    let filesToDelete: string[] = [];

    if (filename) {

      if (txtFiles.includes(filename)) {
        filesToDelete = [filename];
      } else {
        logger.warn('Specific file not found', { filename });
        return res.status(404).json({
          error: `File ${filename} not found`,
          deletedFiles: [],
          deletedCount: 0
        });
      }
    } else if (conversationId) {

      filesToDelete = txtFiles.filter(file => file.startsWith(conversationId));
      if (filesToDelete.length === 0) {
        logger.warn('No files found for conversation', { conversationId });
        return res.status(404).json({
          message: `No files found for conversation ${conversationId}`,
          deletedFiles: [],
          deletedCount: 0
        });
      }
    } else {

      filesToDelete = txtFiles;
    }

    const deletedFiles: string[] = [];
    const failedFiles: FailedFile[] = [];

    for (const file of filesToDelete) {
      try {
        const filePath = path.join(textsDir, file);
        fs.unlinkSync(filePath);
        deletedFiles.push(file);
        logger.info('File deleted successfully', { file });
      } catch (error: any) {
        logger.error('Failed to delete file', { file, error: error.message });
        failedFiles.push({ file, error: error.message });
      }
    }

    const response: {
      message: string;
      deletedFiles: string[];
      deletedCount: number;
      failedFiles?: FailedFile[];
    } = {
      message: `Successfully deleted ${deletedFiles.length} file(s)`,
      deletedFiles,
      deletedCount: deletedFiles.length
    };

    if (failedFiles.length > 0) {
      response.failedFiles = failedFiles;
      response.message += `. Failed to delete ${failedFiles.length} file(s)`;
    }

    logger.info('Delete operation completed', response);
    return res.status(200).json(response);

  } catch (error: any) {
    logger.error('Error in /delete-texts handler', { error: error.message });
    return res.status(500).json({ error: 'Internal server error', message: error.message });
  }
});

export default router;
